docs/docs/09-rfcs/2023-03-08-wing-preview-environment.md

```diff
@@ -0,0 +1,278 @@
+provide a streamlined process for updating a pull request preview environment, each PR code changes will redeploy the preview environment and also will:
+
+1. provide a real-time build and deployment status updates
+2. each entry point will have a seperated preview environment
+3. ensure a consistent and unique url for each preview environment. The url structure: `https://<gh-repository>-<gh-branch>-<entry-point>-<pr-number>.wingcloud.app`
+4. provide the ability to download the preview environment deployment logs for debugging purposes in case of failure, logs urls structure is: `https://wing.cloud/gh-account/gh-repo/gh-branch/entry-file/logs/`
+
+PR comment example:
+
+| Entry Point     | Status                                                                                   | Preview                                                                          | Updated (UTC)        |
+| --------------- | ---------------------------------------------------------------------------------------- | -------------------------------------------------------------------------------- | -------------------- |
+| main.w          | ✅ Ready ([logs](https://wing.cloud/gh-account/gh-repo/gh-branch/main/logs/))            | [Visit Preview](https:/gh-repository-gh-branch-3312-main.wing.cloud.app)         | Jul 31, 2023 8:01am  |
+| another.main.w  | ✅ Ready ([logs](https://wing.cloud/gh-account/gh-repo/gh-branch/another-main/logs/))    | [Visit Preview](https:/gh-repository-gh-branch-3312-another-main.wing.cloud.app) | Jul 30, 2023 10:01am |
+| failed.main.w   | ❌ Failed ([logs](https://wing.cloud/gh-account/gh-repo/gh-branch/entry-file/logs/))     |                                                                                  | Jul 31, 2023 8:01am  |
+| building.main.w | 🔄 Building ([logs](https://wing.cloud/gh-account/gh-repo/gh-branch/entry-file/logs/))   |                                                                                  | Jul 31, 2023 8:01am  |
+| stale.main.w    | ⏸️ Stale ([lean more](https://wing.cloud/gh-account/gh-repo/gh-branch/entry-file/logs/)) |                                                                                  | Jul 31, 2023 8:01am  |
+
+TBD - tests results per entry point
+
+**Disclaimer**: Developers that have write access and can create PRs but didn't signed-up to Wing Cloud will not be able to create preview environments.
+In this case the above comment will contain a direct link for signing up to Wing Cloud.
+After signing up, the developer will be able to create a preview environment for the PR and one will be created automatically after the next code change.
+
+#### Multiple Entry Points
+
+A developer can configure multiple entry points in a single PR.
+Every `*.main.w` file in the repository will be considered as an entry point and will have a dedicated preview environment.
+All preview environments will be available in the PR comment and will be updated upon code changes.
```

Comment:
What is the experience in the console when a PE is updated? Do you get some popup saying "please wait while redeploying PE..."? Or does it break because the backend container died under the hood?

Reply:
There's no signal of change on the frontend. It'd be a nice feature, though!

The websocket connection will break, so events from backend to frontend will stop working. The rest of the information, which is fetched via HTTP requests, will still work but there will be a drastic change because the environment will be wiped clean (eg, no files in buckets, no events, no logs, etc).

Reply:
added this as a requirement, @yoav-steinberg good point